mantidproject/mantid
# Extend SpecularReflectionPositionCorrect

**Description of work.**

This PR adds several new features to the `SpecularReflectionPositionCorrect` algorithm:
* Support for horizontal reflectometers
* Support for detectors which always face the sample
* Support for negative 2theta
* Support for given pixel position correction (at least in the rotate-around-sample case)
* Support for angle calibration using a direct beam reference
* Support for unnamed sample components

**Report to:** [nobody].

**To test:**

Make sure the unit test data directories are included in MantidPlot's data search directories. Check that the script below runs. It exercises most of the changes introduces in this PR.

```python
import numpy as np
reflected = Load('ILL/D17/317370.nxs')
reflected = ExtractMonitors(reflected, DetectorWorkspace='reflected')

# Set detector centre angle to some value
corrected1 = SpecularReflectionPositionCorrect(
    reflected,
    2.3,
    'RotateAroundSample',
    'detector',
    DetectorFacesSample=True)
# The centre is in the middle of pixels 127 and 128
assert(corrected1.spectrumInfo().twoTheta(127) > np.deg2rad(2.3))
assert(corrected1.spectrumInfo().twoTheta(128) < np.deg2rad(2.3))

# Set the angle of certain pixel (workspace index)
# This should be equivalent to above
corrected2 = SpecularReflectionPositionCorrect(
    reflected,
    2.3,
    'RotateAroundSample',
    'detector',
    DetectorFacesSample=True,
    LinePosition=127.5,
    PixelSize=0.001195)
assert(corrected1.spectrumInfo().twoTheta(0) == corrected2.spectrumInfo().twoTheta(0))
# An arbitrary pixel
corrected3 = SpecularReflectionPositionCorrect(
    reflected,
    2.3,
    'RotateAroundSample',
    'detector',
    DetectorFacesSample=True,
    LinePosition=231.87,
    PixelSize=0.001195)
print(np.rad2deg(corrected3.spectrumInfo().twoTheta(232)))


# Correct angle using a direct beam reference
LoadILLReflectometry('ILL/D17/317369.nxs', OutputWorkspace='direct', OutputBeamPosition='direct_beam_pos')
direct = ExtractMonitors('direct', DetectorWorkspace='direct')
direct_beam_pos = mtd['direct_beam_pos'].cell('PeakCentre', 0)
print(direct_beam_pos)
# The direct beam is not quite at 0 degrees.
print(np.rad2deg(direct.spectrumInfo().twoTheta(int(direct_beam_pos))))
# Reflected detector angle needs to be calibrated accordingly
corrected4 = SpecularReflectionPositionCorrect(
    reflected,
    DetectorCorrectionType='RotateAroundSample',
    DetectorComponentName='detector',
    DetectorFacesSample=True,
    PixelSize=0.001195,
    DirectLinePosition=direct_beam_pos,
    DirectLineWorkspace='direct')
print(np.rad2deg(reflected.spectrumInfo().twoTheta(0)))
print(np.rad2deg(corrected4.spectrumInfo().twoTheta(0)))
```

Fixes #22734.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
